src/software/ai/evaluation/calc_best_shot.cpp

```diff
@@ -114,4 +122,147 @@ namespace Evaluation
                                .abs();
         return shot.getOpenAngle().toDegrees() / goal_angle.toDegrees();
     }
+
+    Shot calcMostOpenDirection(Point origin, Segment segment,
+                               std::vector<Robot> robot_obstacles)
+    {
+        std::vector<Circle> obstacles;
+
+        for (Robot robo : robot_obstacles)
+        {
+            obstacles.push_back(Circle(robo.position(), ROBOT_MAX_RADIUS_METERS));
+        }
+
+        return calcMostOpenDirection(origin, segment, obstacles);
+    }
+    Shot calcMostOpenDirection(Point origin, Segment segment,
```

Comment:
This function is pretty long and complicated, so I think it would really benefit from breaking each "state" into its own function. Then the top-level function will read a lot more like pseudocode, like

```
projectObstacles();
createBlockingSegments();
combineBlockingSegments();
findOpenSegments();
```

Reply:
:+1:

Reply:
bumping this comment

Reply:
I broke it up into 2 more functions.

projectCirclesOntoSegment() and getEmptySpaceWithinParentSegment()

Feel free to comment on the names. 

I only did the 2 functions because projectObstacles() and createBlockingSegments() are essentially the same thing